kernel/mempool.c

```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <ksched.h>
+#include <wait_q.h>
+#include <init.h>
+#include <string.h>
+
+/* Linker-defined symbols bound the static pool structs */
+extern struct k_mem_pool _k_mem_pool_list_start[];
+extern struct k_mem_pool _k_mem_pool_list_end[];
+
+s64_t _tick_get(void);
+
+static struct k_mem_pool *get_pool(int id)
+{
+	return &_k_mem_pool_list_start[id];
+}
+
+static int pool_id(struct k_mem_pool *pool)
+{
+	return pool - &_k_mem_pool_list_start[0];
+}
+
+static void *block_ptr(struct k_mem_pool *p, size_t lsz, int block)
+{
+	return p->buf + lsz * block;
+}
+
+static int block_num(struct k_mem_pool *p, void *block, int sz)
+{
+	return (block - p->buf) / sz;
+}
+
+static bool level_empty(struct k_mem_pool *p, int l)
+{
+	return sys_dlist_is_empty(&p->levels[l].free_list);
+}
+
+/* Places a 32 bit output pointer in word, and an integer bit index
+ * within that word as the return value
+ */
+static int get_bit_ptr(struct k_mem_pool *p, int level, int bn, u32_t **word)
+{
+	u32_t *bitarray = level <= p->max_inline_level ?
+		&p->levels[level].bits : p->levels[level].bits_p;
+
+	*word = &bitarray[bn / 32];
+
+	return bn & 0x1f;
+}
+
+static void set_free_bit(struct k_mem_pool *p, int level, int bn)
+{
+	u32_t *word;
+	int bit = get_bit_ptr(p, level, bn, &word);
+
+	*word |= (1<<bit);
```

Comment:
strictly speaking, I think our coding style encourages spaces around the "<<" operator